### PR TITLE
(closes #1456) Updates to PSyAD test harness to allow for precision.

### LINKED ISSUE #1456: [psyad] use consistent precision within generated test-harness code.
Currently the test harness generated as part of the PSyAD example sometimes reports success and sometimes failure.

There are two reasons for this:

 - [x] The fields are declared as REAL but the inner products are calculated using double precision. (We should use the same precision as the fields.)
 - [x] The mechanism used to compare the results does not allow for machine precision and should be replaced by code based upon that supplied by the Met Office.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Review by rupertford — CHANGES_REQUESTED** [review]
Good stuff. Suggestions are minor and mainly about the user guide. There is one suggestion to add another exception for the main psyad script. All code is pycodestyle and pylint clean apart from one highlighted case. All examples run, including the harness example - whoop!

File: doc/psyad/user_guide/implementation.rst [hunk 1]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
              and after it is modified from within active
              statements. This case is not checked in this version, see
              issue #1458.
+
+Test Harness
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:315]: Done.

File: doc/psyad/user_guide/implementation.rst [hunk 2]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+++++++++++++
+
+The code generated for the test harness must perform the following steps:
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:318]: Done.

File: doc/psyad/user_guide/implementation.rst [hunk 3]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+2) Call the tangent-linear kernel;
+3) Compute the inner product of the results of the kernel;
+4) Call the adjoint of the TL kernel, passing in the outputs of the TL
+   kernel call;
+5) Compute the inner product of the results of the adjoint kernel with
+   the original inputs to the TL kernel;
+6) Compare the two inner products for equality, allowing for machine
+   precision.
+
+Some of these steps are described in more detail below.
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:330]: Done.

File: doc/psyad/user_guide/implementation.rst [hunk 4]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+
+All arguments to the TL kernel are initialised with pseudo-random numbers
+in the interval :math:`[0.0,1.0]` using the Fortran `random_number` intrinsic
+function.
+
+Inner products
+--------------
+
+The precision of the variables used to accumulate the inner-product values
+is set to match that of the active variables in the supplied TL kernel.
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:343]: Done.

File: doc/psyad/user_guide/implementation.rst [hunk 5]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+
+For simplicity, when computing the inner product in steps 3) and 5), *all*
+kernel arguments are included (since those that are only read by a kernel
+will remain constant for both the TL and adjoint kernel calls). It may be
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:347]: I've made it explicit.

File: doc/psyad/user_guide/implementation.rst [hunk 6]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+that this will require refinement in future, e.g. if there are kernels
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:348]: Re-worded although I think dofmap arrays will be fine. I'm sure there are other things that might not be though.

File: src/psyclone/psyad/tl2ad.py [hunk 7]
```diff
@@ -45,20 +45,17 @@
...
 from psyclone.psyir.frontend.fortran import FortranReader
 from psyclone.psyir.nodes import Routine, Assignment, Reference, Literal, \
-    Call, Container, BinaryOperation, UnaryOperation, Return, IfBlock, \
+    Call, Container, BinaryOperation, UnaryOperation, IfBlock, \
     CodeBlock, FileContainer, ArrayReference, Range
-from psyclone.psyir.symbols import SymbolTable, ImportInterface, \
+from psyclone.psyir.symbols import SymbolTable, ImportInterface, Symbol, \
     ContainerSymbol, ScalarType, ArrayType, RoutineSymbol, DataSymbol, \
-    INTEGER_TYPE, REAL_DOUBLE_TYPE
+    INTEGER_TYPE, REAL_TYPE
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:52]: You are correct.

File: src/psyclone/psyad/tl2ad.py [hunk 8]
```diff
@@ -155,6 +153,45 @@ def _find_container(psyir):
...
+    intrinsic = None
+    for var in active_variables:
+        sym = kernel.symbol_table.lookup(var)
+        if precision is None:
+            precision = sym.datatype.precision
+            intrinsic = sym.datatype.intrinsic
+        else:
+            if (precision != sym.datatype.precision or
+                    intrinsic != sym.datatype.intrinsic):
+                raise NotImplementedError(
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:186]: Done.

File: src/psyclone/psyad/tl2ad.py [hunk 9]
```diff
@@ -155,6 +153,45 @@ def _find_container(psyir):
...
+                    "Found active variables of different datatype: '{0}' is "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:187]: Good point, done.

File: src/psyclone/psyad/tl2ad.py [hunk 10]
```diff
@@ -155,6 +153,45 @@ def _find_container(psyir):
                               "Containers. This is not supported.")
 
 
+def _get_active_variable_datatype(kernel, active_variables):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:156]: Done.

File: src/psyclone/psyad/tl2ad.py [hunk 11]
```diff
@@ -437,50 +495,35 @@ def generate_adjoint_test(tl_psyir, ad_psyir):
         statements.append(CodeBlock([ptree], CodeBlock.Structure.STATEMENT))
         statements.append(
             Assignment.create(Reference(sym_copy), Reference(sym)))
+    statements[0].preceding_comment = ("Initialise the kernel arguments and "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:498]: Yes, this is becoming a common problem (e.g. for the OpenCL code generation). I chose to do it this way as it keeps it language agnostic and saves us having to worry about how to distribute and build any supporting code.

File: src/psyclone/tests/psyad/main_test.py [hunk 12]
```diff
@@ -64,26 +64,39 @@
...
+  ! Test the inner-product values for equality, allowing for the precision \
+of the active variables
+  MachineTol = SPACING(MAX(ABS(inner1), ABS(inner2)))
+  relative_diff = ABS(inner1 - inner2) / MachineTol
+  if (relative_diff < overall_tolerance) then
+    WRITE(*, *) 'Test of adjoint of ''kern'' PASSED: ', inner1, inner2, \
+relative_diff
+  else
+    WRITE(*, *) 'Test of adjoint of ''kern'' FAILED: ', inner1, inner2, \
+relative_diff
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyad/main_test.py:98]: Done.

**Comment by arporter** [comment]
This is just waiting on Tim's answers to the questions about the implementation of the check for equality of the two inner products.

**Comment by rupertford** [comment]
> This is just waiting on Tim's answers to the questions about the implementation of the check for equality of the two inner products.

I'm happy to review the other changes in the meantime if that would be useful.

**Comment by arporter** [comment]
That's a good idea; let's be asynchronous :-)

File: doc/psyad/user_guide/implementation.rst [hunk 13]
```diff
@@ -311,3 +311,59 @@ individually to each statement and the resultant PSyIR returned.
...
+
+Performing the comparison of the two inner products while allowing for
+machine precision is implemented as follows:
+
+1) Find the smallest possible difference that can be represented by
+   calling the Fortran `spacing` intrinsic on the largest of the two
+   inner-product values;
+
+2) Compute the *relative* difference between the two values by dividing
+   their absolute difference by this spacing;
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst:374]: Updated now.

File: doc/psyad/user_guide/implementation.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/psyad/user_guide/implementation.rst]: I'm afraid I still don't understand the point you are making. Is it that passive variables remain constant and therefore can be included in the inner product computation without affecting the correctness test?

File: src/psyclone/psyad/tl2ad.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyad/tl2ad.py:186]: I suggest merging this with `KeyError` i.e. `(KeyError, NotImplementedError)`

File: src/psyclone/tests/psyad/main_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyad/main_test.py:98]: Could you please explain how the error comes about in `test_main_not_implemented_error` in a sentence at the end of the docstring (see previous test `test_main_keyerror` as an example).

**Review by rupertford — CHANGES_REQUESTED** [review]
Just a few small remaining requests. All else is OK. Please also bring branch up-to-date.

File: doc/psyad/user_guide/implementation.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/psyad/user_guide/implementation.rst]: Yes. I've tried re-writing with your words :-)

File: src/psyclone/psyad/tl2ad.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:186]: Oh yes, that's better.

File: src/psyclone/tests/psyad/main_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyad/main_test.py:98]: Done.

**Comment by arporter** [comment]
CI-happiness permitting, this is ready for another look.

File: doc/psyad/user_guide/implementation.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/psyad/user_guide/implementation.rst]: Thanks, that sounds much better (to me).

**Review by rupertford — APPROVED** [review]
All issues have been addressed. Approving.

**Pull request merged**